### ISSUE TITLE
Rough fine-tuning guidance

### ISSUE THREAD

**Issue by RonanKMcGovern** [metavoiceio/metavoice-src#70]
I know the repo ReadMe says "soon", but would it be possible to give some very rough advice on how to fine-tune to improve on the voice's match with a custom speaker?

I guess the demo is just extracting embeddings from bria.mp3, but I'd like to go one step further to get a better voice match. Thanks.

**Comment by RonanKMcGovern** [comment]
Many thanks @vatsalaggarwal .

Re embeddings, I used a 90 s recording of my voice. I used an mp3 file, so perhaps I could have done better there, but I'm Irish so maybe that was the issue in getting a good match. I'll try out an American person's embeddings instead and see.

Yeah, I think I've got it on training. I guess the first stage produces encodec tokens? What do I use to decode those?

BTW, what model are you using for diffusion in the third stage, I don't see that on the model card on HF, but maybe I glanced over it? Thanks

**Comment by vatsalaggarwal** [comment]
Yeah, and a 44.1/48khz wav or 256kbps mp3 works better… but, it’s unlikely the model would be able to produce an Irish accent regardless in one-shot, at least we didn’t focus on it for this release. 

The first stage produces the first two codebooks of the encodec RVQ.

you can decode these using the second-stage + MBD + deepfilternet. 

We are using multi band diffusion from audio craft. 

More details are available in https://github.com/metavoiceio/metavoice-src?tab=readme-ov-file#architecture

**Comment by deeprobo-dev** [comment]
Hi if I want to fine tune for different language other than english then which stage is most suitable for finetuning, can you please share some insights on that? Also, I working on robotics application and would love to see faster inference with as much lesser vram as possible as I have to run it in a nvidia agx or orion.

**Comment by hrachkovinovoto** [comment]
> Hi if I want to fine tune for different language other than English then which stage is most suitable for fine tuning, can you please share some insights on that? Also, I working on robotics application and would love to see faster inference with as much lesser vram as possible as I have to run it in a nvidia agx or orion.
> 
> Thanks for the amazing project

+1, It'd be great if it manages to run on 8 GB VRAM at the very least.

**Comment by vatsalaggarwal** [comment]
> Hi if I want to fine tune for different language other than English then which stage is most suitable for fine tuning, can you please share some insights on that? Also, I working on robotics application and would love to see faster inference with as much lesser vram as possible as I have to run it in a nvidia agx or orion.

I think you'd need to finetune both the first stage for sure, and I'm not sure if second stage needs to be finetuned, that's something you'd have to check.

Thanks for the request regarding faster inference and on-device.

**Comment by deeprobo-dev** [comment]
> > Hi if I want to fine tune for different language other than English then which stage is most suitable for fine tuning, can you please share some insights on that? Also, I working on robotics application and would love to see faster inference with as much lesser vram as possible as I have to run it in a nvidia agx or orion.
> 
> I think you'd need to finetune both the first stage for sure, and I'm not sure if second stage needs to be finetuned, that's something you'd have to check.
> 
> Thanks for the request regarding faster inference and on-device! How much VRAM and compute (TFLOPs) are available?

Thanks for your insights regarding fine tuning. I will give it a try. For NVIDIA JETSON AGX XAVIER its of 32 TFLOPS and for NVIDIA JETSON AGX ORIN is of 275 TFLOPS.

**Comment by maepopi** [comment]
Hello! If I understood correctly from this thread, it will soon be possible to finetune a model both with a model checkpoint and a LoRA with 12Gb VRAM (and possibly 8gb?)? How large the audio dataset should be for each? 

thank you, this is very exciting!

**Comment by RonanKMcGovern** [comment]
> Hey! Yes, when we released the repo, we thought it'd be the next thing we release, but we've reprioritised, so are busy with other things, and we've not actively started working on releasing finetuning support.

Just to add to this. I wouldn't underestimate how valuable it would be to release fine-tuning in a simple way. I'm not aware of frameworks that accurately can fine-tune for a specific voice. I'd be keen to make a video on this on youtube.com/@trelisresearch if this becomes possible - it allows for things like making your own audio books.

**Comment by maepopi** [comment]
> Just to add to this. I wouldn't underestimate how valuable it would be to release fine-tuning in a simple way. I'm not aware of frameworks that accurately can fine-tune for a specific voice. I'd be keen to make a video on this on youtube.com/@TrelisResearch if this becomes possible - it allows for things like making your own audio books.

If I might, have you checked out [this tool?](https://git.ecker.tech/mrq/ai-voice-cloning) it's based off Tortoise-TTS and it's really good. I've been playing around with it for months and come up with pretty good models. I don't think it supports LoRA's though, which is why I'm very interested in the present repo which seems to provide with a slightly better base model than Tortoise (but this would need to be tested further, it's just an impression for now).

**Comment by danablend** [comment]
> Yeah, and a 44.1/48khz wav or 256kbps mp3 works better… but, it’s unlikely the model would be able to produce an Irish accent regardless in one-shot, at least we didn’t focus on it for this release.
> 
> The first stage produces the first two codebooks of the encodec RVQ.
> 
> you can decode these using the second-stage + MBD + deepfilternet.
> 
> We are using multi band diffusion from audio craft.
> 
> More details are available in https://github.com/metavoiceio/metavoice-src?tab=readme-ov-file#architecture and [#70 (comment)](https://github.com/metavoiceio/metavoice-src/issues/70#issuecomment-1957337895)

Hey! Thanks very much for your insight.

I'm in the midst of attempting to implement fine tuning, and I've gotten a very simple script to train but I could only get it to work by sequentially iterating over each data entry in a batch.

Would you happen to have an idea of how it might look adding batch inference support, so it can train on batches at a time during training / fine tuning?

Again, very much appreciate your work!

**Comment by vatsalaggarwal** [comment]
@danablend can you push your code to PR? I can have a look

**Comment by danablend** [comment]
> @danablend can you push your code to PR? I can have a look

Hey @vatsalaggarwal, I realized that I had made a mistake and needed to build more code to make the training work.

I've spent a few hours working on it, but I get OOM when attempting to train the model with gradients enabled for stage 1 (but fine for remaining stages) on an A10G (16 GB VRAM), so I don't know yet if it works to push to the codebase. How many GB VRAM did you find you needed to train the model?

I'll play around with it more and see if I can prepare something useful and clean for you as a base to work off if that would be helpful?

**Comment by vatsalaggarwal** [comment]
Really hard to know where the issue is without seeing the code! I think it should be possible to finetune on a 16GB GPU but depends on your config (batch size, optimiser choice, etc)...

If it's in a state where I'll be able to run it, that's ideal, but also happy to look it in it's current state and give pointers to speed you up even if it's in a super dirty state.

**Comment by danablend** [comment]
> Really hard to know where the issue is without seeing the code! I think it should be possible to finetune on a 16GB GPU but depends on your config (batch size, optimiser choice, etc)...
> 
> If it's in a state where I'll be able to run it, that's ideal, but also happy to look it in it's current state and give pointers to speed you up even if it's in a super dirty state.

I'll open a PR shortly

**Comment by danablend** [comment]
> Really hard to know where the issue is without seeing the code! I think it should be possible to finetune on a 16GB GPU but depends on your config (batch size, optimiser choice, etc)...
> 
> If it's in a state where I'll be able to run it, that's ideal, but also happy to look at it in it's current state (even if it's in a super dirty state) and give pointers to speed you up.

Just added a PR draft (https://github.com/metavoiceio/metavoice-src/pull/82).
